---
title: "kap1_dashboard"
output:
  flexdashboard::flex_dashboard
---

```{r}
hmchimpCorr <- read.csv("hmchimp_results/hmCorrResult.csv")
chimpCorr <- read.csv("hmchimp_results/chimpCorrResult.csv")

hmoranCorr <- read.csv("hmoran_results/hmCorrResult.csv")
oranCorr <- read.csv("hmoran_results/oranCorrResult.csv")

select_neg <- function(df) {
    df <- df[complete.cases(df), ]
    dfSig <- df[df$padj < 0.05, ]
    dfSig <- dfSig[dfSig$coef < 0, ]
    dfSig
}

hmchimpCorrSig <- select_neg(hmchimpCorr)
chimpCorrSig <- select_neg(chimpCorr)

hmoranCorrSig <- select_neg(hmoranCorr)
oranCorrSig <- select_neg(oranCorr)

# intersect overlapped KRAB-ZNFs from both species
hmCorr <- hmchimpCorrSig[hmchimpCorrSig$geneName %in% hmoranCorrSig$geneName,]
```

```{r drawing heatmap}
library(ggplot2)
library(heatmaply)
library(reshape2)

pair_heatmap <- function(df, top=40, chr=NULL, lowCol="#d2e1b1", highCol="#bf8924"){

    if(!is.null(chr)){
        # filter chromosome
        df <- df[df$chromosome==chr, ]
    }
    
    # select top genes
    df <- df[order(df$Count, decreasing=TRUE), ][seq_len(top), ]
    df <- df[, c("ZNF_ms", "te_name", "Count")]
    dft <- dcast(df, te_name~ZNF_ms, value.var="Count", sum)
    rownames(dft) <- dft$te_name
    dft <- dft[-1]
    dft[is.na(dft)] <- 0
    
    #figPath <- file.path(figDir, figName)
    
    
    heatmaply(as.matrix(dft), 
              colors = c(low=lowCol, high=highCol),
              xlab="KRAB-ZNFs",
              ylab="TEs"
    )
    

}
```



Intro {data-navmenu="Home"}
=================================================
We use `TEKRABber` to analyze the differntially expressed orthologs and also 
analyze the correlations of KRAB-ZNFs and TEs. The data comes from human, 
chimpanzee, and orangutans B-cells RNA-seq data (3 biological replicates/ 
each species). We comparing three species using human as references. 



Human (with Chimpanzee) {data-navmenu="Correlation Tables"}
=================================================
### Human Correlation Table (with Chimpanzee)
This table shows the correlation results from human which comparing with 
Chimpanzee. There are total 101 rows of data and are filtered with padj less 
than 0.05 and coefficients less than -0.7.
```{r}
DT::datatable(
    hmchimpCorrSig,
    fillContainer = FALSE,
    options = list(bPaginate=TRUE, pageLength=5),
    filter = "top"
)

```



Human (with Orangutan) {data-navmenu="Correlation Tables"}
=================================================
### Human Correlation Table (with Orangutan)
This table shows the correlation results from human which comparing with 
Orangutans. There are total 16 rows of data and are filtered with padj less 
than 0.05 and coefficients less than -0.7.
```{r}
DT::datatable(
    hmoranCorrSig,
    fillContainer = FALSE,
    options = list(bPaginate=TRUE, pageLength=5),
    filter = "top"
)
```



Human {data-navmenu="Correlation Tables"}
================================================
### Human Correlation Table (intersect from chimpanzee and orangutan group)
This table shows the intersect of humans KRAB-ZNFs and TEs from comparing human 
with chimpanzee and human with orangutans. The data has been filtered with padj 
less than 0.05 and coefficients less than -0.7. There are total 13 rows of data.

```{r}
DT::datatable(
    hmCorr,
    fillContainer = FALSE,
    options = list(bPaginate=TRUE, pageLength=5),
    filter = "top"
)
```




Chimpanzee {data-navmenu="Correlation Tables"}
=================================================
This table shows chimpanzees KRAB-ZNFs and TEs result with padj less than 0.05
and coefficient less than -0.7.
```{r}
DT::datatable(
    chimpCorrSig,
    fillContainer = FALSE,
    options = list(bPaginate=TRUE, pageLength=5),
    filter = "top"
)
```

Orangutans {data-navmenu="Correlation Tables"}
=================================================
This table shows orangutans KRAB-ZNFs and TEs result with padj less than 0.05
and coefficient less than -0.7.
```{r}
DT::datatable(
    oranCorrSig,
    fillContainer = FALSE,
    options = list(bPaginate=TRUE, pageLength=5),
    filter = "top"
)
```

Intro {data-navmenu="Overlapped"}
=================================================
### Intro
The input tables are corrdinates information of KRAB-ZNFs and TEs based on 
different method. Two of them are matrices built using **JASPAR Core 
nonredundant vertebrates-2020 collection**, **Zinc finger motifs for human in 
the JASPAR Collection Unvalidated** and **collection of motifs assembled by 
Dogan et al**. Upon these four data are intersect by RNA-seq data result which 
include significantly correlated KRAB-ZNFs and TEs from TEKRABber's result.

```{r}
# here we try to select the KRAB-ZNFs result
```

Human {data-navmenu="Overlapped"}
=================================================

Column
-------------------------------------------------
### Clustering pair counts in Human (JASPAR & other database) {data-padding=50}  
We visualize the pair counts comparing motifs from KRAB-ZNFs and TEs from JASPAR 
and TEKRABber results. Here we only show the top 40 counts.  
```{r}
hmdbTable <- read.csv("intersect/hm_db_repeat.csv")

pair_heatmap(hmdbTable)
```

### Clustering pair counts in Human (ChIP-Seq) {data-padding=50}  
```{r}
hmkatTable <- read.csv("intersect/hm_kat_repeat.csv")
pair_heatmap(hmkatTable)
```

Human (Ch19) {data-navmenu="Overlapped"}
=================================================

Column
-------------------------------------------------
### Clustering pair counts in Human (JASPAR & other database) {data-padding=50}  
We visualize the pair counts comparing motifs from KRAB-ZNFs and TEs from JASPAR 
and TEKRABber results. Here we only show the top 40 counts.  
```{r}
hmdbTable <- read.csv("intersect/hm_db_repeat.csv")

pair_heatmap(hmdbTable, chr="chr19")
```

### Clustering pair counts in Human (ChIP-Seq) {data-padding=50}  
```{r}
hmkatTable <- read.csv("intersect/hm_kat_repeat.csv")
pair_heatmap(hmkatTable, chr="chr19")
```

Chimpanzee {data-navmenu="Overlapped"}
=================================================

Column
-------------------------------------------------
### Clustering pair counts in Chimpanzee (JASPAR & other database) {data-padding=50}  
We visualize the pair counts comparing motifs from KRAB-ZNFs and TEs from JASPAR 
and TEKRABber results. Here we only show the top 40 counts.  
```{r}
chimpdbTable <- read.csv("intersect/chimp_db_repeat.csv")

pair_heatmap(chimpdbTable)
```

### Clustering pair counts in Chimpanzee (ChIP-Seq) {data-padding=50}  
```{r}
chimpkatTable <- read.csv("intersect/chimp_kat_repeat.csv")
pair_heatmap(chimpkatTable)
```


Orangutans {data-navmenu="Overlapped"}
=================================================

Column
-------------------------------------------------
### Clustering pair counts in Orangutans (JASPAR & other database) {data-padding=50}  
We visualize the pair counts comparing motifs from KRAB-ZNFs and TEs from JASPAR 
and TEKRABber results. Here we only show the top 40 counts.  
```{r}
orandbTable <- read.csv("intersect/oran_db_repeat.csv")

pair_heatmap(orandbTable)
```

### Clustering pair counts in Orangutans (ChIP-Seq) {data-padding=50}  
```{r}
orankatTable <- read.csv("intersect/oran_kat_repeat.csv")
pair_heatmap(orankatTable)
```

Intro {data-navmenu="Bipartite"}
=================================================
### Intro
In this section, we use **bipartite** graph to visualize selected ZNF_ms and 
TEs.

The input tables are corrdinates information of KRAB-ZNFs and TEs based on 
different method. Two of them are matrices built using **JASPAR Core 
nonredundant vertebrates-2020 collection**, **Zinc finger motifs for human in 
the JASPAR Collection Unvalidated** and **collection of motifs assembled by 
Dogan et al**. Upon these four data are intersect by RNA-seq data result which 
include significantly correlated KRAB-ZNFs and TEs from TEKRABber's result.

```{r}
library(igraph)
library(plyr)

# load hm file
load("intersect/db_tight.rda")
load("intersect/katja_tight.rda")
hm_db <- df_db_tight_split$hm
hm_kat <- kat_tight_split$hm

hm_db_select <- hm_db[hm_db$chromosome=="chr19",c("ZNF_ms", "te_name")]
hm_kat_select <- hm_kat[,c("ZNF_ms", "te_name")]


# select top40
counts <- ddply(hm_db_select, .(hm_db_select$ZNF_ms, hm_db_select$te_name), nrow)
names(counts) <- c("ZNF_ms", "te_name", "Count")
counts <- counts[order(counts$Count, decreasing = TRUE),]
```

Human parallel (JASPAR) {data-navmenu="Bipartite"}
================================================
### Top 40 pairs Human (JASPAR & other database)
```{r}
g <- graph.data.frame(counts[c(1:40), ], directed = FALSE)
V(g)$type <- bipartite_mapping(g)$type

V(g)$color <- ifelse(V(g)$type, "lightgreen", "salmon")
V(g)$shape <- ifelse(V(g)$type, "circle", "square")
E(g)$color <- "lightgray"

plot(g, layout=layout.bipartite, vertex.label.cex=0.6, vertex.label.color = "black")
```

Human bipartite (JASPAR) {data-navmenu="Bipartite"}
==================================================
### Top 40 pairs Human (JASPAR & other database)
```{r}
plot(g, vertex.label.cex = 0.8, vertex.label.color="black")
```

